pkg/sync/syncer.go

```diff
@@ -1425,6 +1445,538 @@ func (s *syncer) syncGrantsForResource(ctx context.Context, resourceID *v2.Resou
+		if entitlementGrantPageToken == "" {
+			break
+		}
+	}
+	return grantsForEnts, nil
+}
+
+func (s *syncer) listExternalResourceTypes(ctx context.Context) ([]*v2.ResourceType, error) {
+	resourceTypes := make([]*v2.ResourceType, 0)
+	rtPageToken := ""
+	for {
+		resourceTypesResp, err := s.externalResourceReader.ListResourceTypes(ctx, &v2.ResourceTypesServiceListResourceTypesRequest{
+			PageToken: rtPageToken,
+		})
+		if err != nil {
+			return nil, err
+		}
+		resourceTypes = append(resourceTypes, resourceTypesResp.List...)
+		rtPageToken = resourceTypesResp.NextPageToken
+		if rtPageToken == "" {
+			break
+		}
+	}
+	return resourceTypes, nil
+}
+
+func (s *syncer) listAllGrants(ctx context.Context) ([]*v2.Grant, error) {
+	grants := make([]*v2.Grant, 0)
+	pageToken := ""
+	for {
+		grantsResp, err := s.store.ListGrants(ctx, &v2.GrantsServiceListGrantsRequest{
+			PageToken: pageToken,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		grants = append(grants, grantsResp.List...)
+		pageToken = grantsResp.NextPageToken
+		if pageToken == "" {
+			break
+		}
+	}
+	return grants, nil
+}
+
+// TODO(lauren) Split up principals by type (user, group)
+func (s *syncer) processGrantsWithExternalPrincipals(ctx context.Context, principals []*v2.Resource) error {
+	ctx, span := tracer.Start(ctx, "processGrantsWithExternalPrincipals")
+	defer span.End()
+
+	if !s.state.HasExternalResourcesGrants() {
+		return nil
+	}
+
+	grants, err := s.listAllGrants(ctx)
+	if err != nil {
+		return err
+	}
+	expandedGrants := make([]*v2.Grant, 0)
+	for _, grant := range grants {
+		annos := annotations.Annotations(grant.Annotations)
+		matchResourceMatchAllAnno, err := GetExternalResourceMatchAllAnnotation(annos)
+		if err != nil {
+			return err
+		}
+		if matchResourceMatchAllAnno != nil {
+			for _, principal := range principals {
+				rAnnos := annotations.Annotations(principal.GetAnnotations())
+				batonID := &v2.BatonID{}
+				if !rAnnos.Contains(batonID) {
+					continue
+				}
+				switch matchResourceMatchAllAnno.ResourceType {
+				case v2.ExternalResourceType_EXTERNAL_RESOURCE_TYPE_USER:
+					userTrait := &v2.UserTrait{}
+					ok, err := rAnnos.Pick(userTrait)
+					if err != nil {
+						return err
+					}
+					if !ok {
+						continue
+					}
+				case v2.ExternalResourceType_EXTERNAL_RESOURCE_TYPE_GROUP:
+					groupTrait := &v2.GroupTrait{}
+					ok, err := rAnnos.Pick(groupTrait)
+					if err != nil {
+						return err
+					}
+					if !ok {
+						continue
+					}
+				default:
+					return errors.New("unexpected external resource type")
+				}
+
+				newGrant := &v2.Grant{
+					Entitlement: grant.Entitlement,
+					Principal:   principal,
+					Id:          batonGrant.NewGrantID(principal, grant.Entitlement),
+					Sources:     grant.Sources,
+					Annotations: grant.Annotations,
+				}
+				expandedGrants = append(expandedGrants, newGrant)
+			}
+			err = s.store.DeleteGrant(ctx, grant.Id)
```

Comment:
We're deleting the original grant before we're adding the new ones, so if the syncer dies at just the wrong time, we won't know about the original grants when we resume. I think we could collect the grants to delete and do that after the call to PutGrants(...).